pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -13,217 +13,184 @@
-    budget : Dict
-        The total budget.
-
-    Returns
-    -------
-    Dict
-        A dictionary with channels as keys and their respective contributions as values.
-        The key 'total' contains the total expected contribution across all channels.
-
-    Raises
-    ------
-    ValueError
-        If the specified `method` is not recognized.
-    """
+    The goal of this optimization is to maximize the total expected response
+    by allocating the given budget across different marketing channels. The
+    optimization is performed using the Sequential Least Squares Quadratic
+    Programming (SLSQP) method, which is a gradient-based optimization algorithm
+    suitable for solving constrained optimization problems.
 
-    total_expected_contribution = 0.0
-    contributions = {}
-
-    for channel, channe_budget in budget.items():
-        if method == "michaelis-menten":
-            L, k = parameters[channel]
-            contributions[channel] = michaelis_menten(channe_budget, L, k)
-
-        elif method == "sigmoid":
-            alpha, lam = parameters[channel]
-            contributions[channel] = sigmoid_saturation(channe_budget, alpha, lam)
-
-        else:
-            raise ValueError("`method` must be either 'michaelis-menten' or 'sigmoid'.")
-
-        total_expected_contribution += contributions[channel]
-
-    contributions["total"] = total_expected_contribution
-
-    return contributions
-
-
-def objective_distribution(
-    x: list[float],
-    method: str,
-    channels: list[str],
-    parameters: dict[str, tuple[float, float]],
-) -> float:
-    """
-    Compute the total contribution for a given budget distribution.
-
-    This function calculates the negative sum of contributions for a proposed budget
-    distribution using the Michaelis-Menten model. This value will be minimized in
-    the optimization process to maximize the total expected contribution.
+    For more information on the SLSQP algorithm, refer to the documentation:
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html
 
     Parameters
     ----------
-    x : List of float
-        The proposed budget distribution across channels.
-    channels : List of str
-        The List of channels for which the budget is being optimized.
-    parameters : Dict
-        Michaelis-Menten parameters for each channel as described in `calculate_expected_contribution`.
-
-    Returns
-    -------
-    float
-        Negative of the total expected contribution for the given budget distribution.
+    adstock : AdstockTransformation
+        The adstock class.
+    saturation : SaturationTransformation
+        The saturation class.
+    num_days : int
+        The number of days.
+    parameters : dict
+        A dictionary of parameters for each channel.
+    adstock_first : bool, optional
+        Whether to apply adstock transformation first or saturation transformation first.
+        Default is True.
```

Comment:
Missing 

```
  Returns
  -------
  ...
```

Reply:
Shall we use the same types as the `__init__` method? `parameters: dict[str, dict[str, dict[str, float]]]`